src/components/controls/controls.js

```diff
@@ -0,0 +1,220 @@
+var registerControls = require('../../core/controls').registerControls;
+var THREE = require('../../lib/three');
+
+var MAX_DELTA = 0.2; // ms
+var PI_2 = Math.PI / 2;
+
+/**
+ * Controls component.
+ *
+ * Receives input events from device-specific control components, and applies movement and rotation
+ * to the element accordingly.
+ */
+module.exports.Component = registerControls('controls', {
+  dependencies: ['velocity'],
+
+  schema: {
+    enabled: { default: true },
+
+    flyingEnabled: { default: false },
+
+    position: {
+      default: ['hmd-controls', 'gamepad-controls', 'keyboard-controls', 'touch-controls']
+    },
+    positionEnabled: { default: true },
+    positionEasing: { default: 15 }, // m/s2
+    positionAcceleration: { default: 65 }, // m/s2
+
+    rotation: { default: ['hmd-controls', 'gamepad-controls', 'mouse-controls'] },
+    rotationEnabled: { default: true }
+  },
+
+  init: function () {
+    // Movement
+    this.velocity = new THREE.Vector3();
+
+    // Rotation
+    this.pitch = new THREE.Object3D();
+    this.yaw = new THREE.Object3D();
+    this.yaw.position.y = 10;
+    this.yaw.add(this.pitch);
+    this.heading = new THREE.Euler(0, 0, 0, 'YXZ');
+
+    if (this.el.sceneEl.hasLoaded) {
+      this.validateControls();
+    } else {
+      this.el.sceneEl.addEventListener('loaded', this.validateControls.bind(this));
+    }
+  },
+
+  update: function () {
+    // Re-validate if scene has loaded. If not, listener was already bound in init().
+    if (this.el.sceneEl.hasLoaded) {
+      this.validateControls();
+    }
+  },
+
+  tick: function (t, dt) {
+    if (isNaN(dt)) { return; }
+
+    // Update rotation.
+    if (this.data.rotationEnabled) {
+      this.updateRotation(dt);
+    }
+
+    // Update velocity. If FPS is too low, reset.
+    if (this.data.positionEnabled && dt / 1000 > MAX_DELTA) {
+      this.velocity.set(0, 0, 0);
+      this.el.setAttribute('velocity', this.velocity);
+    } else if (this.data.positionEnabled) {
+      this.updateVelocity(dt);
+    }
+  },
+
+  /**
+   * Updates rotation based on input from the active rotation component, if any.
+   * @param  {number} dt
+   */
+  updateRotation: function (dt) {
+    var control, dRotation;
+
+    control = this.getActiveRotationControls();
+    if (control && control.getRotationDelta) {
+      dRotation = control.getRotationDelta(dt);
+      this.yaw.rotation.y -= dRotation.x;
+      this.pitch.rotation.x -= dRotation.y;
+      this.pitch.rotation.x = Math.max(-PI_2, Math.min(PI_2, this.pitch.rotation.x));
+      this.el.setAttribute('rotation', {
+        x: THREE.Math.radToDeg(this.pitch.rotation.x),
+        y: THREE.Math.radToDeg(this.yaw.rotation.y),
+        z: 0
```

Comment:
For like Vive controllers, I see you'd use `getRotation`.

But just want to make it flexible in case you want to, say, click an object to select it, and then scroll the mousewheel to roll it.